### ISSUE TITLE
Generated strings in macros results in DuplicatedLiterals false alarms

### ISSUE THREAD

**Issue by tomassdepakin** [SonarOpenCommunity/sonar-cxx#1147]
Hi!

We use stringizing operator in our code, and scanning it results in triggering StringLiteralDuplicated rule.
https://msdn.microsoft.com/en-us/library/7e3a913x.aspx

How can we avoid triggering this rule in such case?

**Comment by guwirth** [comment]
@tomassdepakin could you provide an concrete sample and the error message please.

**Comment by tomassdepakin** [comment]
@guwirth 
Yep. We have few macro definitions:
1: `#define __FATAL_ASSERT(expression, explanation)  
do { if ((expression) || (!Common::Diagnostics::Detail::DoAssert(#expression, (explanation), __FILE__, __LINE__, false))); else __debugbreak(); } while (false)`
2: `#define WG_FATAL(expression)                       __FATAL_ASSERT(expression, "")`
3: `#define WG_FATAL_MAIN_THREAD()                     WG_FATAL(Common::Diagnostics::MainThreadCheck())`

Then a lot of our code use this macro:
`WG_FATAL_MAIN_THREAD();`

Preprocessor resolve it like:
`do { if ((Common::Diagnostics::MainThreadCheck()) || (!Common::Diagnostics::Detail::DoAssert("Common::Diagnostics::MainThreadCheck()", (""), "src\\auth\\authentication_manager.cpp", 633, false))); else __debugbreak(); } while (false);`

And finally after scanning with sonarqube we have an issue:
![wgfatal](https://cloud.githubusercontent.com/assets/7347796/26392062/e3997300-406d-11e7-945c-a4559d1789e9.png)

**Comment by guwirth** [comment]
@tomassdepakin thanks for the sample. This is a false positive I marked the issue as error.

The string in your sample is a generated string of the preprocessor. The parser in the second step does not detect that string is from a macro.

Think solution is to change the check to create no warning in case of "generated code".

**Comment by Bertk** [comment]
@guwirth A small enhancement of the checkers ```visitNode()``` might solve the issue.

```java
  @Override
  public void visitNode(AstNode node) {
    if (node.is(CxxGrammarImpl.LITERAL) && (!node.getToken().isGeneratedCode())) {
      visitOccurence(node.getTokenOriginalValue(), node.getTokenLine());
    }
  }

```
